### ISSUE TITLE
database connexion error after one night

### ISSUE THREAD

**Issue by stboiss** [stevespringett/dependency-track#131]
I start dependency-track 3.0.2 in docker with a MySQL database.
after one night, i have error on GUI : 
![screenshot 2018-04-12 08-37-52 001](https://user-images.githubusercontent.com/24654609/38660267-ec427886-3e2c-11e8-9eb3-abde06130963.png)
and this error in server.log :
```
2018-04-12 06:33:12,033 [] WARN [org.eclipse.jetty.server.HttpChannel] /api/v1/project
javax.servlet.ServletException: javax.jdo.JDODataStoreException: Could not retrieve transaction read-only status from server
NestedThrowables:
java.sql.SQLException: Could not retrieve transaction read-only status from server
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:432)
        at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:370)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:389)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:342)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:229)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:860)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1650)
        at alpine.filters.ContentSecurityPolicyFilter.doFilter(ContentSecurityPolicyFilter.java:223)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
        at alpine.filters.ClickjackingFilter.doFilter(ClickjackingFilter.java:91)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
        at alpine.filters.WhitelistUrlFilter.doFilter(WhitelistUrlFilter.java:113)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1629)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:190)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:530)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:347)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:256)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
        at java.lang.Thread.run(Thread.java:748)
Caused by: javax.jdo.JDODataStoreException: Could not retrieve transaction read-only status from server
        at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:542)
        at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:456)
        at org.datanucleus.api.jdo.JDOQuery.execute(JDOQuery.java:290)
        at alpine.persistence.AlpineQueryManager.hasPermission(AlpineQueryManager.java:477)
        at alpine.filters.AuthorizationFilter.filter(AuthorizationFilter.java:92)
        at org.glassfish.jersey.server.ContainerFilteringStage.apply(ContainerFilteringStage.java:132)
        at org.glassfish.jersey.server.ContainerFilteringStage.apply(ContainerFilteringStage.java:68)
        at org.glassfish.jersey.process.internal.Stages.process(Stages.java:197)
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:269)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:272)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:268)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:316)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:298)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:268)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:289)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:256)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:703)
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:416)
        ... 40 common frames omitted
Caused by: java.sql.SQLException: Could not retrieve transaction read-only status from server
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:965)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:898)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:887)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:861)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:878)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:874)
        at com.mysql.jdbc.ConnectionImpl.isReadOnly(ConnectionImpl.java:3546)
        at com.mysql.jdbc.ConnectionImpl.isReadOnly(ConnectionImpl.java:3514)
        at org.apache.commons.dbcp2.DelegatingConnection.isReadOnly(DelegatingConnection.java:466)
        at org.apache.commons.dbcp2.DelegatingConnection.isReadOnly(DelegatingConnection.java:466)
        at org.datanucleus.store.rdbms.ConnectionFactoryImpl$ManagedConnectionImpl.getConnection(ConnectionFactoryImpl.java:422)
        at org.datanucleus.store.rdbms.SQLController.getStatementForQuery(SQLController.java:307)
        at org.datanucleus.store.rdbms.query.RDBMSQueryUtils.getPreparedStatementForQuery(RDBMSQueryUtils.java:224)
        at org.datanucleus.store.rdbms.query.JDOQLQuery.performExecute(JDOQLQuery.java:623)
        at org.datanucleus.store.query.Query.executeQuery(Query.java:1966)
        at org.datanucleus.store.query.Query.executeWithArray(Query.java:1855)
        at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:433)
        ... 56 common frames omitted
Caused by: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: The last packet successfully received from the server was 56,217,630 milliseconds ago.  The last packet sent successfully to the server was 56,217,630 milliseconds ago. is longer than the server configured value of 'wait_timeout'. You should consider either expiring and/or testing connection validity before use in your application, increasing the server configured values for client timeouts, or using the Connector/J connection property 'autoReconnect=true' to avoid this problem.
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
        at com.mysql.jdbc.SQLError.createCommunicationsException(SQLError.java:990)
        at com.mysql.jdbc.MysqlIO.send(MysqlIO.java:3749)
        at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2512)
        at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2683)
        at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2482)
        at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2440)
        at com.mysql.jdbc.StatementImpl.executeQuery(StatementImpl.java:1381)
        at com.mysql.jdbc.ConnectionImpl.isReadOnly(ConnectionImpl.java:3539)
        ... 66 common frames omitted
Caused by: java.net.SocketException: Broken pipe (Write failed)
        at java.net.SocketOutputStream.socketWrite0(Native Method)
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111)
        at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
        at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
        at com.mysql.jdbc.MysqlIO.send(MysqlIO.java:3731)
        ... 72 common frames omitted
```

in application.properties, i have : 
```
alpine.database.url=jdbc:mysql://mysql:3306/dtrack?autoReconnect=true&useSSL=false
alpine.database.driver.path=/data/mysql-connector-java-5.1.46-bin.jar
```

in net , i found this : 
https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-usagenotes-troubleshooting.html#qandaitem-14-1-12

**Comment by stevespringett** [comment]
Thank you for reporting this. 👍 

On the MySQL Docker image I use, the default wait_timeout is 28800 seconds (8 hours). As a temporary measure, you can try to increase the value.

I'm planning on swapping connection pool libraries in 3.1 by moving from DBCP2 to HikariCP to increase performance. I was in the middle of testing this when this ticket was created.

`SET GLOBAL wait_timeout = 86400;`

HikariCP is much more graceful than DBCP2 when it comes to reestablishing lost connections. DBCP2 will hand off a closed connection to the application without checking first. HikariCP will check before handing out dead connections and attempt to reestablish one before doing so.

I've replicated with issue using DBCP2 and have confirmed that HikariCP fixes the issue reported. I plan on committing the changes to the trunk and backporting the changes to the 3.0.x branch today. I've got one other major bug to address after which I'll push out 3.0.3.

**Comment by stevespringett** [comment]
This should be resolved in v3.0.3 released today.
